src/core/app.lua

```diff
@@ -444,25 +442,46 @@ function report_apps ()
    end
 end
 
+local next_optimize
 local maxtraces = 100000 -- XXX informal upper bound
 
 -- Check all LuaJIT traces for potential inefficiencies.
 --
 -- Heuristic: If a trace has a loop but spends less than 1/3 of its
 -- time in the looping part then it should be flushed and re-recorded.
 function introspective_optimize ()
+   if next_optimize == nil then
+      next_optimize = monotonic_now
+   elseif monotonic_now >= next_optimize then
+      audit_traceprofile()
+      next_optimize = monotonic_now + 1.0
+   end
+end
+   
+function audit_traceprofile ()
+   local nonloop_total = 0
+   local loop_total = 0
+   local ffi_total = 0
    for tr = 1, maxtraces do
       local info = jitutil.traceinfo(tr)
       -- Just consider looping root traces
       if info and info.mcloop ~= 0 then
-         local nonloop, loop = traceprofile.tracestats(tr)
-         if nonloop > loop*2 then
-            print("flushing "..tr)
-            -- More than twice as much time in the non-loop? Suspicious.
-            jit.flush(tr)
+         local nonloop, loop, ffi = traceprofile.tracestats(tr)
+         if (loop+nonloop > 100) and (nonloop > loop*2) then
+            -- Suspicious trace: has used a significant amount of CPU
+            -- and less than a third of its time spent in the loop.
+            print(("[flushing trace %d: low loop intensity (%d%%)]"):format(
+                  tr, 100*loop/(loop+nonloop)))
```

Comment:
Thanks! I had missed that & could well be the reason that my early tests are showing slow traces being detected but not corrected :-).